04-cs-fundamentals/internship/hash-tables.md

```diff
@@ -0,0 +1,319 @@
+    - Once we have the list of letters and their number of occurances, we can use this to build the output string.
+
+  Because looking up an items in a hash table is so fast, hash tables turn out to be excellent data structures.
+
+</details>
+<br />
+
+#### Exercise
+
+With a neighbor examine the problem below and do the following:
+
+1.  Write down 2-3 clarifying questions.
+1.  Write down assumptions you will make for each clarifying question.
+1.  Generate a fresh set of 2-3 sample input-output looking for **edge cases**.
+1.  Break down the problem into at least 2 subproblems, more if you can.
+1.  Write out a pseudocode solution.
+
+**Problem Description**
+
+Given a string, sort it in decreasing order based on the frequency of characters.
+
+**Example 1:**
+
+```
+Input:
+"tree"
+
+Output:
+"eert"
+
+Explanation:
+'e' appears twice while 'r' and 't' both appear once.
+So 'e' must appear before both 'r' and 't'. Therefore "eetr" is also a valid answer.
+Example 2:
+
+Input:
+"cccaaa"
+
+Output:
+"cccaaa"
+```
+
+**Explanation:**
+Both 'c' and 'a' appear three times, so "aaaccc" is also a valid answer.
+Note that "cacaca" is incorrect, as the same characters must be together.
+
+**Example 3:**
+
+```
+Input:
+"Aabb"
+
+Output:
+"bbAa"
+```
+
+**Explanation:**
+
+"bbaA" is also a valid answer, but "Aabb" is incorrect.
+Note that 'A' and 'a' are treated as two different characters.
+
+You can find a [solution in the example code folder](example-code/character_frequency.rb).
+
+## Array, Linked List, Binary Search Tree or Hash Table
+
+Hash tables are execellent for looking up data from a key.  Binary search trees are excellent data structures for maintaining items in order while linked lists provide quick access to the front and rear and maintain a specific order.
+
+So, which would you use in these situations?
+
+**Question 1:**
+
+<details>
+  <summary>A list of candidate records that need to be stored so that you can find the max or min test scores</summary>
+  Hash tables are not good for maintaining ordered data.  Instead a binary search tree, or sorted array or linked list (sorted by test scores) would serve better.
+</details>
+
+<!-- source: http://www.cs.cmu.edu/~guna/15-123S11/Lectures/Lecture17.pdf -->
+<br />
+
+**Question 2:**
+
+<details>
+  <summary>Fedex needs to provide quick access for customers to check the status of their packages.  Quick lookup time using distinct tracking numbers is important.</summary>
+  Since the tracking numbers are distinct and order is not important, a hash table makes the most sense because of it's quick lookup time.
+</details>
+
+<br />
+
+**Question 3:**
+
+<details>
+  <summary>Postgres needs to keep track of fields in a table to output sorted data.  Items will be regularly inserted and deleted from the table.</summary>
+  Since order is important, maintaining a set of binary search trees would make a great deal of sense.
```

Comment:
It might be worth adding a note about how Postgres "uses a similar structure", but not actually a BST.